### DISCUSSION TITLE
Exception "Not a TestFixture, SetUpFixture or TestSuite, but ParameterizedFixture"

### DISCUSSION THREAD

**Discussion by tmx7777** [nunit/nunit3-vs-adapter#867] [Q&A] [answered]
After upgrading from 3.17.0 to 4.0.0 I'm getting the following exception when running tests within the VS test runner:

> Not a TestFixture, SetUpFixture or TestSuite, but ParameterizedFixture

The class looks basically like this:

```csharp
[TestFixture(typeof(SomeClassA))]
[TestFixture(typeof(SomeClassB))]
[TestFixture(typeof(SomeClassC))]
internal class CommonTests<T> : TestFixtureBase
{
	// ...
}
```
`TestFixtureBase` is a base class of mine. But I doubt this could be problem.

Any help what the problem might be? 
What additional information could I provide to help investigate the issue?

Thank you.

**Comment by OsirisTerje** [comment, marked as answer]
Can you provide a small repro for this?   
It is the internal engine adapter that parses the XML from the engine that is being "surprised" by something here. 
What you can do is to turn on the dump features in the runsettings file and upload the results there.  That will probably give a good hint.    See https://docs.nunit.org/articles/vs-test-adapter/Tips-And-Tricks.html#dumpxmltestdiscovery-and-dumpxmltestresults

**Reply by OsirisTerje** [reply]
I simplified the code above just to see if it was related to the multiple testfixture setups there (which it needs ofc since it is parametrized), and it works as it should:
![image](https://user-images.githubusercontent.com/203432/121409076-5cbf8c80-c961-11eb-967c-9a7fc7570b3f.png)

So what kind of attribute does the TestFixtureBase have ?

**Reply by OsirisTerje** [reply]
I tried to add some more bells and whistles here, but it still works:
![image](https://user-images.githubusercontent.com/203432/121409598-fab35700-c961-11eb-83f3-dbdf1e4d9718.png)

**Reply by OsirisTerje** [reply]
You can modify this repro code yourselves to make it fail, see https://github.com/nunit/nunit3-vs-adapter.issues/tree/master/Issue867

**Reply by OsirisTerje** [reply]
Also, if you need to be unblocked for production use, you can set the Discovery method to Legacy, see https://docs.nunit.org/articles/vs-test-adapter/Tips-And-Tricks.html#discoverymethod

**Reply by tmx7777** [reply]
Thanks for the quick feedback. 
(BTW, this is not an urgent issue for me.)

I played around with my project. Turns out that there is a SetupFixture class within the same namespace. 
I removed all code from my test and all code from the SetupFixture, but the error persists.
However, if I completely remove the SetupFixture, the tests run successfully.

I could set up a proper repro within the next couple of days, if necessary.
But maybe the following code would already be helpful to you.

`SetupFixture.cs`
```csharp
using NUnit.Framework;

namespace Company.Project.Common.Tests
{
    [SetUpFixture]
    public sealed class SetupFixture
    {
        [OneTimeSetUp]
        public void OneTimeSetup()
        {
        }

        [OneTimeTearDown]
        public void OneTimeTeardown()
        {
        }
    }
}
```

`MyTests.cs`
```csharp
using NUnit.Framework;

namespace Company.Project.Common.Tests
{
    [TestFixture(1)]
    [TestFixture(2)]
    internal class MyTests
    {
        public MyTests(int n)
        {
        }

        [Test]
        public void SomeTestMethod()
        {
        }
    }
}
```

**Reply by OsirisTerje** [reply]
Thanks!  I'll check it out.  We know there still are certain patterns of usage that are not covered by the new parser, so everything in this space is appreciated :-)

**Reply by tbutler-qontigo** [reply]
Hi  @OsirisTerje We are hitting a similar problem with [TestFixture("someparam")] - are you guys any closer to a solution?
I can't use a .runsettings file as no control over passing that file to the test runner in a CI environment.
All works fine with 3.17.0
Thanks

**Reply by OsirisTerje** [reply]
You can try the adapter alpha that is enclosed with this issue:  https://github.com/nunit/nunit3-vs-adapter/issues/884 , link to zipped package:  https://github.com/nunit/nunit3-vs-adapter/files/7214073/NUnit3TestAdapter.4.1.0-alpha01.zip

**Reply by OsirisTerje** [reply]
The issue is solved, so it will be part of the 4.1 release

**Reply by tbutler-qontigo** [reply]
That alpha works thanks (though we can't use alpha code in deployed environments).  Do you have a timescale for 4.1?

**Reply by OsirisTerje** [reply]
There is one outstanding issue for 4.1, but the idea is to get it out mid-October.

**Reply by OsirisTerje** [reply]
Parametrized fixture is fixed in 4.1 and upcoming 4.2 will have the fix for generic fixtures.

**Reply by tmx7777** [reply]
Thanks for the update. Unfortunately I'm not able to use 4.0/4.1 for a whole other reason, right now. I'm using a test playlist with some filter settings and tests are executed even though they are supposed to be filtered out. I will try to isolate this with a small repro. Might take a while. Would it be okay to report this issue in the Q&A forum or what would be the proper channel (sorry for my ignorance)?

**Reply by OsirisTerje** [reply]
@tmx7777  Can you have  a look at issue #941 .  This looks like the same as you mention last here.  There is a dev version of the adapter enclosed there.  Can you verify if that one solves your issue ?
